File: src/services/douyu.ts
import fetch, { Response as FetchResponse } from 'node-fetch';
import crypto from 'crypto';
import { JSDOM } from 'jsdom';
import { Agent as HttpsAgent } from 'https';

interface StreamResponse {
  data?: {
    rtmp_url?: string;
    rtmp_live?: string;
  };
  error?: number;
  msg?: string;
}

const agent = new HttpsAgent({
  family: 4,
  keepAlive: true,
  rejectUnauthorized: false
});

class DouYu {
  private did: string;
  private rid: string;
  private readonly maxRetries = 3;
  private readonly retryDelay = 1000;

  constructor(rid: string) {
    this.did = '10000000000000000000000000001501';
    this.rid = rid;
  }

  private md5(data: string): string {
    return crypto.createHash('md5').update(data).digest('hex');
  }

  private async fetchWithRetry(url: string, options: any, retries = this.maxRetries): Promise<FetchResponse> {
    try {
      const response = await fetch(url, options);
      
      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }
      
      return response;
    } catch (error) {
      if (retries > 0) {
        console.log(`重试获取直播流 (剩余重试次数: ${retries - 1})`);
        await new Promise(resolve => setTimeout(resolve, this.retryDelay));
        return this.fetchWithRetry(url, options, retries - 1);
      }
      throw error;
    }
  }

  private async getPcJs(cdn: string = 'tct', rate: number = 0): Promise<string> {
    try {
      const response = await this.fetchWithRetry(`https://www.douyu.com/${this.rid}`, {
        headers: {
          'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
          'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8',
          'Accept-Language': 'zh-CN,zh;q=0.9',
          'Cache-Control': 'no-cache',
          'Pragma': 'no-cache'
        },
        agent: agent
      });

      const text = await response.text();

      // 检查页面是否包含错误信息
      if (text.includes('房间未找到') || text.includes('房间已关闭')) {
        console.log('房间未开播或不存在');
        return '';
      }

      // 提取JS函数
      const funcReg = /(vdwdae325w_64we[\s\S]*function ub98484234[\s\S]*?)function/;
      const funcMatch = text.match(funcReg);
      if (!funcMatch) {
        console.log('无法找到必要的JS函数');
        return '';
      }

      const evalReg = /eval\(.*?\);/g;
      const funcUb9 = funcMatch[1].replace(evalReg, 'strc;');

      // 创建虚拟DOM环境
      const dom = new JSDOM('', { runScripts: 'dangerously' });
      const window = dom.window;
      
      window.eval(funcUb9);
      const result = window.eval('ub98484234()') as string;

      const vReg = /v=(\d+)/;
      const vMatch = result.match(vReg);
      if (!vMatch) throw new Error('v parameter not found');

      const v = vMatch[1];
      const t10 = Math.floor(Date.now() / 1000).toString();

      const rb = this.md5(`${this.rid}${this.did}${t10}${v}`);

      let funcSign = result
        .replace('return rt;})', 'return rt;}')
        .replace('(function (', 'function sign(')
        .replace('CryptoJS.MD5(cb).toString()', `"${rb}"`);

      window.eval(funcSign);
      const params = window.eval(`sign("${this.rid}", "${this.did}", "${t10}");`) as string;

      const finalParams = `${params}&cdn=${cdn}&rate=${rate}`;

      // 获取真实URL
      const apiResponse = await this.fetchWithRetry(`https://www.douyu.com/lapi/live/getH5Play/${this.rid}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Origin': 'https://www.douyu.com',
          'Referer': `https://www.douyu.com/${this.rid}`,
          'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
          'Cache-Control': 'no-cache',
          'Pragma': 'no-cache'
        },
        body: finalParams,
        agent: agent
      });

      const json = await apiResponse.json() as StreamResponse;

      if (json.error && json.error !== 0) {
        console.log('获取直播流失败:', json.msg);
        return '';
      }

      if (!json.data) {
        console.log('响应中缺少数据字段');
        return '';
      }

      const { rtmp_url, rtmp_live } = json.data;
      if (!rtmp_url || !rtmp_live) {
        console.log('缺少流地址组件');
        return '';
      }

      // 尝试不同的 CDN
      const cdns = ['tct', 'ws-h5', 'tx-h5'];
      if (!cdns.includes(cdn)) {
        return this.getPcJs(cdns[0], rate);
      }

      return `${rtmp_url}/${rtmp_live}`;

    } catch (error) {
      console.log('获取直播流出错:', error);
      return '';
    }
  }

  public async getStreamUrl(): Promise<string> {
    try {
      const url = await this.getPcJs();
      if (!url) {
        console.log('房间未开播或获取流地址失败');
        return '';
      }
      return url;
    } catch (error) {
      console.log('获取直播流失败:', error);
      return '';
    }
  }
}

export async function getStreamUrl(roomId: string): Promise<string> {
  const douyu = new DouYu(roomId);
  return await douyu.getStreamUrl();
}

export default DouYu; 

File: src-tauri/src/douyu.rs
use md5::Digest;
use quick_js::{Context, JsValue};
use regex::Regex;
use std::time::{SystemTime, UNIX_EPOCH};
use isahc::{prelude::*, HttpClient, Request, http};
use isahc::config::{RedirectPolicy, Configurable};
use percent_encoding::{percent_encode, NON_ALPHANUMERIC};

struct DouYu {
    did: String,
    rid: String,
    client: HttpClient,
}

impl DouYu {
    async fn new(rid: &str) -> Result<Self, Box<dyn std::error::Error>> {
        println!("Creating new DouYu instance for room {}", rid);
        let client = HttpClient::builder()
            .redirect_policy(RedirectPolicy::Follow)
            .default_header("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36")
            .build()?;

        Ok(Self {
            did: "10000000000000000000000000001501".to_string(),
            rid: rid.to_string(),
            client,
        })
    }

    fn md5(data: &str) -> String {
        let mut hasher = md5::Md5::new();
        hasher.update(data.as_bytes());
        format!("{:x}", hasher.finalize())
    }

    async fn get_pc_js(&self, cdn: &str, rate: i32) -> Result<String, Box<dyn std::error::Error>> {
        println!("Fetching webpage for room {}", self.rid);
        // 获取PC网页内容
        let request = Request::builder()
            .method(http::Method::GET)
            .uri(&format!("https://www.douyu.com/{}", self.rid))
            .header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8")
            .header("Accept-Language", "zh-CN,zh;q=0.9")
            .header("Connection", "keep-alive")
            .body(())?;

        let text = self.client.send(request)?
            .text()?;

        println!("Extracting JS function from webpage");
        // 提取JS函数
        let re = Regex::new(r"(vdwdae325w_64we[\s\S]*function ub98484234[\s\S]*?)function")?;
        let result = re.captures(&text)
            .ok_or("Cannot find js function")?
            .get(1)
            .ok_or("No capture group")?
            .as_str();

        let re_eval = Regex::new(r"eval.*?;\}")?;
        let func_ub9 = re_eval.replace_all(result, "strc;}");

        println!("Executing JS to get parameters");
        // 执行JS获取参数
        let context = Context::new()?;
        context.eval(&func_ub9)?;
        let js_result = context.eval("ub98484234()")?;
        let res = match js_result {
            JsValue::String(s) => s,
            _ => return Err("JavaScript didn't return a string".into())
        };

        println!("Extracting v parameter");
        // 提取v参数
        let re = Regex::new(r"v=(\d+)")?;
        let v = re.captures(&res)
            .ok_or("v parameter not found")?
            .get(1)
            .ok_or("No capture group")?
            .as_str();

        let t10 = SystemTime::now()
            .duration_since(UNIX_EPOCH)?
            .as_secs()
            .to_string();

        let rb = Self::md5(&format!("{}{}{}{}", self.rid, self.did, &t10, v));

        println!("Building sign function");
        // 构造签名函数
        let mut func_sign = res.replace("return rt;})", "return rt;}");
        func_sign = func_sign.replace("(function (", "function sign(");
        func_sign = func_sign.replace("CryptoJS.MD5(cb).toString()", &format!("\"{}\"", rb));

        context.eval(&func_sign)?;
        let sign_call = format!(
            "sign(\"{}\", \"{}\", \"{}\");",
            self.rid, self.did, t10
        );
        let js_params = context.eval(&sign_call)?;
        let mut params = match js_params {
            JsValue::String(s) => s,
            _ => return Err("JavaScript didn't return a string".into())
        };

        params.push_str(&format!("&cdn={}&rate={}", cdn, rate));

        println!("Fetching stream URL for room {}", self.rid);
        // 获取真实URL
        let url = format!("https://www.douyu.com/lapi/live/getH5Play/{}", self.rid);
        let request = Request::builder()
            .method(http::Method::POST)
            .uri(&url)
            .header("Content-Type", "application/x-www-form-urlencoded")
            .header("Origin", "https://www.douyu.com")
            .header("Referer", format!("https://www.douyu.com/{}", self.rid))
            .body(params)?;

        let mut response = self.client.send(request)?;
        println!("API response status: {}", response.status());
        
        let json: serde_json::Value = response.json()?;

        let data = json["data"].as_object().ok_or("No data field in response")?;
        let rtmp_url = data["rtmp_url"].as_str().ok_or("No rtmp_url field")?;
        let rtmp_live = data["rtmp_live"].as_str().ok_or("No rtmp_live field")?;

        let final_url = format!("{}/{}", rtmp_url, rtmp_live);
        println!("Successfully got stream URL for room {}", self.rid);
        
        Ok(final_url)
    }

    pub async fn get_real_url(&self) -> Result<String, Box<dyn std::error::Error>> {
        self.get_pc_js("ws-h5", 0).await
    }
}

pub async fn get_stream_url(room_id: &str) -> Result<String, Box<dyn std::error::Error>> {
    println!("Starting to fetch stream URL for room {}", room_id);
    let douyu = DouYu::new(room_id).await?;
    let url = douyu.get_real_url().await?;
    Ok(url)  // 直接返回实际的流地址
}

// 添加搜索主播的函数
pub async fn search_anchor(keyword: &str) -> Result<String, Box<dyn std::error::Error>> {
    let client = HttpClient::builder()
        .redirect_policy(RedirectPolicy::Follow)
        .default_header("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36")
        .build()?;

    let mut hasher = md5::Md5::new();
    hasher.update(SystemTime::now()
        .duration_since(UNIX_EPOCH)?
        .as_nanos()
        .to_string());
    let did = format!("{:x}", hasher.finalize());

    let url = format!(
        "https://www.douyu.com/japi/search/api/searchUser?kw={}&page=1&pageSize=20&filterType=0",
        percent_encode(keyword.as_bytes(), NON_ALPHANUMERIC)
    );

    let request = Request::builder()
        .method(http::Method::GET)
        .uri(url)
        .header("Referer", "https://www.douyu.com/search/")
        .header("Cookie", format!("dy_did={}; acf_did={}", did, did))
        .body(())?;

    let mut response = client.send(request)?;
    let text = response.text()?;
    
    Ok(text)
} 